### PR TITLE
Add getSelfUser to ShardManager.

### PR DESCRIPTION
**Pull request by Blayyke**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

There are several guidelines you should follow in order for your
  Pull Request to be merged.

- [X] I have checked the PRs for upcoming features/bug fixes.
- [X] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Description

This PR allows getting the SelfUser from ShardManager, similar to ShardManager#getApplicationInfo.

### CODE REVIEW

**Comment by kantenkugel** [comment]
This isn't as easy as you make it look like. While your code does return a `SelfUser` instance, its just ONE of many (different) `SelfUser` instances because every shard has its own `SelfUser`.

I personally also don't see what the benefit of this would be opposed to just getting the `SelfUser` from the JDA insance you are currently working in (as every JDA class has `getJDA()`)

**Comment by Blayyke** [comment]
Ah, didn't realize shards had different SelfUsers. Assumed they were all the same provided the user wasn't different on any shards. A use-case I had for this was updating bot stats on the two of the bot lists, where i needed the bot's ID.

**Comment by kantenkugel** [comment]
Well yea... they would equal() to each other and have same id, but they are different instances of SelfUser. Maybe Im just making it looks worse as it is. I'll wait what other devs say to this. might just be accepted, as long as docs note that you shouldn't use Shard specific methods like getJDA() on the SelfUser

**Comment by MinnDevelopment** [comment]
Can I get a user-case for this?

File: src/main/java/net/dv8tion/jda/bot/sharding/ShardManager.java [hunk 1]
```diff
@@ -104,6 +105,22 @@ default int getShardsTotal()
...
+     *         If there is no running shard
+     *
+     * @return The {@link net.dv8tion.jda.core.entities.SelfUser} for the currently logged in account.
+     */
+    default SelfUser getSelfUser()
+    {
+        return this.getShardCache().stream()
+                .findAny()
+                .orElseThrow(() -> new IllegalStateException("no active shards"))
+                .getSelfUser()
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/bot/sharding/ShardManager.java]: I somehow don't believe this would compile

Comment (Blayyke) [reply src/main/java/net/dv8tion/jda/bot/sharding/ShardManager.java]: Don't know how I missed that one, will fix.

**Comment by Blayyke** [comment]
One use-case was in my message above

**Comment by MinnDevelopment** [comment]
current way:
```java
updateStats(selfUser);

public void updateStats(SelfUser user) {
    ShardManager manager = user.getJDA().getShardManager();
}
```
vs suggested way:
```java
updateStats(manager);

public void updateStats(ShardManager manager) {
     SelfUser self = manager.getSelfUser();
}
```

I see no benefit here?

**Comment by Blayyke** [comment]
I send a stat update on bot startup, so I didn't really have some proper way of getting the SelfUser. Also thought it may fit because getApplicationInfo exists.

**Comment by WalshyDev** [comment]
Maybe instead of this just add getId() and getIdLong() for the SelfUser, this would still suite the use-case and mean you don't have to just get the first JDA instance every-time in order to get the ID.

Old way
```java
long id = shardManager.getShards().get(0).getSelfUser().getIdLong()
```

New way
```java
long id = shardManager.getIdLong()
```

I would like to see a getSelfUser though from the first connected shard it can find, the reason being if let's say shard 0 goes down and you're using the bots avatar in embeds then it can just die until the shard reconnects.

Something like
```java
public SelfUser getSelfUser() {
        for (JDA jda : getShardCache())
            if (jda.getStatus() == JDA.Status.CONNECTED)
                return jda.getSelfUser();
        return null;
    }
```

Would be nice but I'm not sure if this would be the best and most reliable way.

**Comment by MinnDevelopment** [comment]
We decided that this is not needed. You should implement this yourself if you need this functionality.

```java
public static SelfUser getSelfUser(ShardManager manager) {
    return manager.getShardCache().stream()
            .filter( (shard) -> shard.getStatus() == JDA.Status.CONNECTED)
            .map(JDA::getSelfUser)
            .findAny().orElseThrow(IllegalStateException::new);
}
```

**Pull request closed without merging**